### ISSUE TITLE
buffered channels

### ISSUE THREAD

**Issue by cdelorme** [cdelorme/level#3]
I had buffers for my channels when writing the second draft of the source, but after tests concluded not to include them.
## test case

I tested without buffers, with buffers the size of the number of workers, twice the number of workers, and five times the number of workers:
- 0
- 1x
- 2x
- 5x

My tests were run on debian linux x64 with 32GB of memory.

I used two data sets:
- single-copy of 12GB of documents
- duplicated copies of those 12GB as a biased data-set to test hash performance

The first allowed me to test a more feasible situation, one where there are likely to be significantly less duplicates (it was an old set of personal documents, so some duplicates were sure to exist).

The second was forcefully biased, to test hash performance and would certainly not represent a normal data-set, nor would it necessarily yield unbiased performance results (more time spent hashing that throws off how long the channels take to process the data)

For all instances I ran the program ten times against the datasets, including summary, verbose logging, and cpu profiling output.
## performance

**Biased Data (tested first):**

Summary output for each indicated 206242 files scanned, and 151674 hashes generated (crc32 and sha256).

In all cases the CPU profile was unvariably similar, with the most time spent in Syscall and sha256 hashing functions.

My tests yielded very similar performance metrics for no-buffer and 2x buffered channels with time, with a slightly more erratic performance without buffers occassionally being .05 seconds slower.

According to the debug output, the spread of work ranged between 800-1300 for each worker.  This range remained the same for nearly all test cases, with the 800-900 & 1200-1300 being outliers in all cases.  As the buffer sizes increased the range become slightly tighter, but the outliers occurred at least once in every test.

The speed of the 1x and 5x buffers was approx. .1 seconds slower.

_Again, because a majority of time spent in these tests was likely processing duplicate hashes the timing is biased._

**Normal Data:**

Summary output revealed 103121 files scanned, with 74270 crc32 hashes generated and 29786 sha256 hashes generated (a great example of crc32 not being as accurate but great as a front, because that's 50,000 less potential sha256 hashes it may have generated).

In this case cpu profiling revealed that we still had a majority of time spend running syscall and sha256, but this time sha256 was significantly less (but still the second highest time-spent function).  _I could conclude from these results that none of the code I wrote in this project is a major bottleneck, **though this does not mean my code cannot be improved.**_

Work distribution ranged between 700-1100, and outliers in the 700 & 1000 ranges had roughly the same rate of occurence throughout the tests.

Execution time ranges:
- 1.27-1.31 seconds w/o buffers
- 1.28-1.31 seconds with 1x buffers
- 1.30-1.36 seconds with 2x buffers
- 1.35-1.42 seconds with 5x buffers
## conclusion

In conclusion I removed channel buffers from my source.  The buffers did not appear to greatly improve channel distribution, and in almost all cases they resulted in lower speed of execution; with the 2x buffer being the exception when working on a moderately large data set and only by .05 seconds.

_This may be different on a single-core system, and if anyone is willing to supply me with some test results I'd be happy to see them._
